### ISSUE TITLE
Grid: encourage the zero-arg constructor pattern and add columns

### ISSUE THREAD

**Issue by mvysny** [mvysny/karibu-dsl#4]
Vaadin 10 Grid sports two constructors - zero-arg and one-arg taking Class in. While the one-arg Class constructor looks handy since it pre-creates columns for all properties, it has the following inherent disadvantages:

- With Vaadin 10, once the column is created, it is no longer possible to change `Renderer` nor `ValueProvider` and thus it's not possible to present the value in other way but by Grid calling `toString()` on the value.
- To change the Renderer/ValueProvider, the column has to be removed and a new one can be created.
- That leads to a spaghetti code where a bunch of columns are removed and re-created.
- Since the order of the column can not be changed programatically, to change the column order you need to remove all columns and re-create them.

The one-arg Class constructor is therefore perfect for but a simplest use cases; a bit more complex use cases quickly lead to spaghetti code. Therefore, we will prefer the zero-arg constructor and we will create the columns by hand. We will offer the `addColumnFor` function which inherently does the right thing:
- it sets the Column key so that the column is back-referenceable
- if sorting is enabled, the sort property and the comparator is set, so that the sorting works both on in-memory and on backend `DataProvider`s
- the header is auto-generated by converting camelCase to Human Friendly
- runs a block which allows further modification to the column.

That leads to the following code:
```kotlin
grid {
    isExpand = true
    addColumnFor(Report::priority)
    addColumnFor(Report::type)
    addColumnFor(Report::summary)
    addColumnFor(Report::assigned, sortable = false, converter = { it?.name })
    addColumnFor(Report::timestamp, converter = { getHumanTimeAgo(it) })
    addColumnFor(Report::reportedTimestamp, converter = { getHumanTimeAgo(it) })
}
```

The column order and the column contents are immediately visible from the code above. Therefore we will consider this practice superior to pre-created columns with Vaadin 10 Grids.

Revisit the Vaadin 8 Grids to adopt similar practice. If this practice is adopted for Vaadin 8 Grids, we need to:
- update all Vaadin 8 tutorials
- update all Vaadin 8 example projects

**Comment by mvysny** [comment]
Vaadin 8 Grid column in fact allows to change the renderer after creation (or even a `ValueProvider`+`Renderer` combo), yet there are other arguments:
- The order of auto-created columns depends on the order of fields in the bean. Changing the order of the properties in the bean would cause the order of the Grid columns to change, and this connection is not apparent
- The magic of auto-generated columns promotes more magic by configuring stuff like column header via annotations on the properties; on the other hand explicit configuration via `addColumnFor` promotes the builder pattern which removes magic and promotes code locality.

**Comment by mvysny** [comment]
Much better:
```kotlin
        personGrid = grid(dataProvider = Person.dataProvider) {
            expandRatio = 1f; setSizeFull()

            // a sample of how to reconfigure a column
            addColumnFor(Person::id) { isSortable = false }
            addColumnFor(Person::name)
            addColumnFor(Person::age)
            addColumnFor(Person::dateOfBirth) {
                setRenderer(LocalDateRenderer())
            }
            addColumnFor(Person::maritalStatus)
            addColumnFor(Person::alive)
            addColumnFor(Person::created) {
                // example of a custom renderer which converts value to a displayable string.
                setRenderer({ it.toString() }, TextRenderer())
            }

            // add additional columns with buttons
            addColumn({ "Show" }, ButtonRenderer<Person>({ event -> PersonView.navigateTo(event.item) }))
            addColumn({ "Edit" }, ButtonRenderer<Person>({ event -> createOrEditPerson(event.item) })).id = "edit"
            addColumn({ "Delete" }, ButtonRenderer<Person>({ event -> deletePerson(event.item.id!!) }))

            // automatically create filters, based on the types of values present in particular columns.
            appendHeaderRow().generateFilterComponents(this, Person::class)
        }
```

**Comment by mvysny** [comment]
Fixed in Karibu-DSL 0.3.9
